game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveSubsRetreatTest.java

```diff
@@ -0,0 +1,195 @@
+
+@ExtendWith(MockitoExtension.class)
+class OffensiveSubsRetreatTest {
+
+  @Mock ExecutionStack executionStack;
+  @Mock IDelegateBridge delegateBridge;
+  @Mock BattleActions battleActions;
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource
+  void retreatHappens(final String displayName, final BattleState battleState) {
+    final OffensiveSubsRetreat offensiveSubsRetreat =
+        new OffensiveSubsRetreat(battleState, battleActions);
+
+    offensiveSubsRetreat.execute(executionStack, delegateBridge);
+
+    verify(battleActions)
+        .queryRetreat(eq(false), eq(MustFightBattle.RetreatType.SUBS), eq(delegateBridge), any());
+  }
+
+  static List<Arguments> retreatHappens() {
+    return List.of(
+        Arguments.of(
+            "Can not submerge but has retreat territories",
+            givenBattleStateBuilder()
+                .attackingUnits(List.of(givenUnitCanEvade()))
+                .gameData(
+                    MockGameData.givenGameData().withTransportCasualtiesRestricted(false).build())
+                .attackerRetreatTerritories(List.of(mock(Territory.class)))
+                .build()),
+        Arguments.of(
+            "Has no retreat territories but can submerge",
+            givenBattleStateBuilder()
+                .attackingUnits(List.of(givenUnitCanEvade()))
+                .gameData(
+                    MockGameData.givenGameData()
+                        .withTransportCasualtiesRestricted(false)
+                        .withSubmersibleSubs(true)
+                        .build())
+                .attackerRetreatTerritories(List.of())
+                .build()),
+        Arguments.of(
+            "Transports with other units on the defense",
+            givenBattleStateBuilder()
+                .attackingUnits(List.of(givenUnitCanEvade()))
+                .gameData(
+                    MockGameData.givenGameData()
+                        .withTransportCasualtiesRestricted(true)
+                        .withSubmersibleSubs(true)
+                        .build())
+                .defendingUnits(List.of(givenUnitTransport(), givenAnyUnit()))
+                .build()));
+  }
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource
+  void retreatDoesNotHappen(final String displayName, final BattleState battleState) {
+    final OffensiveSubsRetreat offensiveSubsRetreat =
+        new OffensiveSubsRetreat(battleState, battleActions);
+
+    offensiveSubsRetreat.execute(executionStack, delegateBridge);
+
+    verify(battleActions, never()).queryRetreat(anyBoolean(), any(), any(), any());
+  }
+
+  static List<Arguments> retreatDoesNotHappen() {
+    return List.of(
+        Arguments.of(
+            "Battle is over",
+            givenBattleStateBuilder()
+                .attackingUnits(List.of(givenUnitCanEvade()))
+                .over(true)
+                .build()),
+        Arguments.of(
+            "Defending Destroyer exists",
```

Comment:
I think I might be missing something. A defending destroyer prevents submerge, why does it prevent retreat?

Reply:
The special sub submerge/retreat step is separate from the main retreat step.  This special step will either allow submerging or retreating depending on a gameData parameter (`getSubmersibleSubs`).